File: GdPicture/WinForm/C#/OCR/Document.cs
using System.Drawing;
using GdPicture14;

namespace Ocr
{
    public class Document
    {
        /// <summary>
        /// The constructor initializes members to their default values.
        /// </summary>
        public Document(GdPictureImaging gdPictureImaging, GdPictureOCR gdPictureOcr)
        {
            _gdPictureImaging = gdPictureImaging;
            _gdPictureOcr = gdPictureOcr;
            ImageId = -1;
            _ocrResultId = "";
        }

        /// <summary>
        /// Load document loads the document.
        /// </summary>
        /// <param name="filePath">The path to the file to load.</param>        
        /// <param name="pdfRasterizationResolution">The resolution for the rasterization of the pdfs.</param>
        /// <returns>true of the load succeeds otherwise returns false.</returns>
        public bool Load(string filePath, float pdfRasterizationResolution)
        {
            // In case the document is a pdf, it is required to rasterize the page.

            if (GdPictureDocumentUtilities.GetDocumentFormat(filePath) == GdPicture14.DocumentFormat.DocumentFormatPDF)
            {
                using (GdPicturePDF gdPicturePdf = new GdPicturePDF())
                {
                    if (gdPicturePdf.LoadFromFile(filePath, false) == GdPictureStatus.OK)
                    {
                        ImageId = gdPicturePdf.RenderPageToGdPictureImageEx(pdfRasterizationResolution, true);
                        gdPicturePdf.CloseDocument();
                    }
                }
            }
            else
            {
                ImageId = _gdPictureImaging.CreateGdPictureImageFromFile(filePath);
            }

            if (ImageId == 0)
            {
                return false;
            }

            // Take into accourant the rotation information stored within the exif tags.

            GdPictureRotateFlipType rotateFlipType = _gdPictureImaging.TagGetExifRotation(ImageId);
            if (rotateFlipType != GdPictureRotateFlipType.GdPictureRotateNoneFlipNone)
            {
                _gdPictureImaging.Rotate(ImageId, (RotateFlipType)rotateFlipType);
                _gdPictureImaging.TagDeleteAll(ImageId);
            }

            return true;
        }

        /// <summary>
        /// Close closes the current document if one.
        /// </summary>
        public void Close()
        {
            DiscardOcrResult();
            if (ImageId != 0)
            {
                _gdPictureImaging.ReleaseGdPictureImage(ImageId);
                ImageId = 0;
            }
        }

        /// <summary>
        /// DiscardOcrResult discards the current ocr result.
        /// </summary>
        public void DiscardOcrResult()
        {
            if (!string.IsNullOrWhiteSpace(_ocrResultId))
            {
                _gdPictureOcr.ReleaseOCRResult(_ocrResultId);
                _ocrResultId = "";
            }
        }

        /// <summary>
        /// The identifier for the current image.
        /// </summary>
        public int ImageId { get; private set; }

        /// <summary>
        /// The identifier for the ocr result.
        /// </summary>
        public string OcrResultId
        {
            get { return _ocrResultId; }
            set
            {
                DiscardOcrResult();
                _ocrResultId = value;
            }
        }

        /// <summary>
        /// The flag indicates whether a document is open or not.
        /// </summary>
        public bool IsOpen
        {
            get { return ImageId > 0; }
        }

        /// <summary>
        /// The flag indicates whether a document has been recognized or not.
        /// </summary>
        public bool HasOcr
        {
            get { return !string.IsNullOrWhiteSpace(_ocrResultId); }
        }

        /// <summary>
        /// The boxes surrounding the blocks.
        /// </summary>
        public Rectangle[] BlocksBoundingBoxes
        {
            get { return HasOcr ? Utilities.GetBlocksBoundingBoxes(_ocrResultId, _gdPictureOcr) : null; }
        }

        /// <summary>
        /// The boxes surrounding the paragraphs.
        /// </summary>
        public Rectangle[] ParagraphsBoundingBoxes
        {
            get { return HasOcr ? Utilities.GetParagraphsBoundingBoxes(_ocrResultId, _gdPictureOcr) : null; }
        }

        /// <summary>
        /// The boxes surrounding the text lines.
        /// </summary>
        public Rectangle[] TextLinesBoundingBoxes
        {
            get { return HasOcr ? Utilities.GetTextLinesBoundingBoxes(_ocrResultId, _gdPictureOcr) : null; }
        }

        /// <summary>
        /// The boxes surrounding the words.
        /// </summary>
        public Rectangle[] WordsBoundingBoxes
        {
            get { return HasOcr ? Utilities.GetWordsBoundingBoxes(_ocrResultId, _gdPictureOcr) : null; }
        }

        /// <summary>
        /// The boxes surrounding the characters.
        /// </summary>
        public Rectangle[] CharactersBoundingBoxes
        {
            get { return HasOcr ? Utilities.GetCharactersBoundingBoxesSimplified(_ocrResultId, _gdPictureOcr) : null; }
        }

        /// <summary>
        /// The last OCR result.
        /// </summary>
        public string OCRResult
        {
            get { return HasOcr ? _gdPictureOcr.GetOCRResultText(_ocrResultId) : null; }
        }

        /// <summary>
        /// GdPictureImaging instance.
        /// </summary>
        private readonly GdPictureImaging _gdPictureImaging;

        /// <summary>
        /// GdPictureOcr instance.
        /// </summary>
        private readonly GdPictureOCR _gdPictureOcr;

        /// <summary>
        /// The identifier for the ocr result.
        /// </summary>
        private string _ocrResultId;
    }
}

File: GdPicture/WinForm/vb.net/OCR/Document.vb
Imports GdPicture14

Namespace Ocr

    Public Class Document

        Public Sub New(ByVal gdPictureImaging As GdPictureImaging, ByVal gdPictureOcr As GdPictureOCR)
            _gdPictureImaging = gdPictureImaging
            _gdPictureOcr = gdPictureOcr
            ImageId = -1
            _ocrResultId = ""
        End Sub

        Public Function Load(ByVal filePath As String, ByVal detectOrientation As Boolean, ByVal deskew As Boolean, ByVal pdfRasterizationResolution As Single) As Boolean
            If GdPictureDocumentUtilities.GetDocumentFormat(filePath) = GdPicture14.DocumentFormat.DocumentFormatPDF Then
                Using gdPicturePdf As GdPicturePDF = New GdPicturePDF()
                    If gdPicturePdf.LoadFromFile(filePath, False) = GdPictureStatus.OK Then
                        ImageId = gdPicturePdf.RenderPageToGdPictureImageEx(pdfRasterizationResolution, True)
                        gdPicturePdf.CloseDocument()
                    End If
                End Using
            Else
                ImageId = _gdPictureImaging.CreateGdPictureImageFromFile(filePath)
            End If

            If ImageId = 0 Then
                Return False
            End If

            Dim rotateFlipType As GdPictureRotateFlipType = _gdPictureImaging.TagGetExifRotation(ImageId)
            If rotateFlipType <> GdPictureRotateFlipType.GdPictureRotateNoneFlipNone Then
                _gdPictureImaging.Rotate(ImageId, CType(rotateFlipType, RotateFlipType))
                _gdPictureImaging.TagDeleteAll(ImageId)
            End If

            If detectOrientation Then
                _gdPictureOcr.SetImage(ImageId)
                Dim orientationAngle As Integer = _gdPictureOcr.GetOrientation()
                If orientationAngle <> 0 Then
                    Dim correctOrientationAngle As Integer = 360 - orientationAngle
                    _gdPictureImaging.RotateAngle(ImageId, CSng(correctOrientationAngle))
                End If
            End If

            If deskew Then
                _gdPictureImaging.AutoDeskew(ImageId)
            End If

            Return True
        End Function

        Public Sub Close()
            DiscardOcrResult()
            If ImageId <> 0 Then
                _gdPictureImaging.ReleaseGdPictureImage(ImageId)
                ImageId = 0
            End If
        End Sub

        Public Sub DiscardOcrResult()
            If Not String.IsNullOrWhiteSpace(_ocrResultId) Then
                _gdPictureOcr.ReleaseOCRResult(_ocrResultId)
                _ocrResultId = ""
            End If
        End Sub

        Public Property ImageId As Integer

        Public Property OcrResultId As String
            Get
                Return _ocrResultId
            End Get

            Set(ByVal value As String)
                DiscardOcrResult()
                _ocrResultId = value
            End Set
        End Property

        Public Property IsOpen As Boolean
            Get
                Return ImageId > 0
            End Get
            Private Set(value As Boolean)
            End Set
        End Property

        Public Property HasOcr As Boolean
            Get
                Return Not String.IsNullOrWhiteSpace(_ocrResultId)
            End Get
            Private Set(value As Boolean)
            End Set
        End Property

        Public Property BlocksBoundingBoxes As Rectangle()
            Get
                Return If(HasOcr, Utilities.GetBlocksBoundingBoxes(_ocrResultId, _gdPictureOcr), Nothing)
            End Get
            Private Set(value As Rectangle())
            End Set
        End Property

        Public Property ParagraphsBoundingBoxes As Rectangle()
            Get
                Return If(HasOcr, Utilities.GetParagraphsBoundingBoxes(_ocrResultId, _gdPictureOcr), Nothing)
            End Get
            Private Set(value As Rectangle())
            End Set
        End Property

        Public Property TextLinesBoundingBoxes As Rectangle()
            Get
                Return If(HasOcr, Utilities.GetTextLinesBoundingBoxes(_ocrResultId, _gdPictureOcr), Nothing)
            End Get
            Private Set(value As Rectangle())
            End Set
        End Property

        Public Property WordsBoundingBoxes As Rectangle()
            Get
                Return If(HasOcr, Utilities.GetWordsBoundingBoxes(_ocrResultId, _gdPictureOcr), Nothing)
            End Get
            Private Set(value As Rectangle())
            End Set
        End Property

        Public Property CharactersBoundingBoxes As Rectangle()
            Get
                Return If(HasOcr, Utilities.GetCharactersBoundingBoxesSimplified(_ocrResultId, _gdPictureOcr), Nothing)
            End Get
            Private Set(value As Rectangle())
            End Set
        End Property

        Public Property OCRResult As String
            Get
                Return If(HasOcr, _gdPictureOcr.GetOCRResultText(_ocrResultId), Nothing)
            End Get
            Private Set(value As String)
            End Set
        End Property

        Private ReadOnly _gdPictureImaging As GdPictureImaging

        Private ReadOnly _gdPictureOcr As GdPictureOCR

        Private _ocrResultId As String
    End Class
End Namespace
